flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ShowDatabasesOperation.java

```diff
@@ -20,26 +20,108 @@
+        }
+    }
+
+    public ShowDatabasesOperation(
+            String preposition,
+            String catalogName,
+            String likeType,
+            String likePattern,
+            boolean notLike) {
+        this.preposition = preposition;
+        this.catalogName = catalogName;
+        if (likeType != null) {
+            this.likeType = LikeType.of(likeType);
+            this.likePattern = requireNonNull(likePattern, "Like pattern must not be null");
+            this.notLike = notLike;
+        } else {
+            this.likeType = null;
+            this.likePattern = null;
+            this.notLike = false;
+        }
+    }
+
     @Override
     public String asSummaryString() {
-        return "SHOW DATABASES";
+        StringBuilder builder = new StringBuilder();
+        builder.append("SHOW DATABASES");
+        if (preposition != null) {
+            builder.append(String.format(" %s %s", preposition, catalogName));
+        }
+        if (likeType != null) {
+            if (notLike) {
+                builder.append(String.format(" NOT %s '%s'", likeType.name(), likePattern));
+            } else {
+                builder.append(String.format(" %s '%s'", likeType.name(), likePattern));
+            }
+        }
+        return builder.toString();
     }
 
     @Override
     public TableResultInternal execute(Context ctx) {
-        String[] databases =
-                ctx.getCatalogManager()
-                        .getCatalogOrThrowException(ctx.getCatalogManager().getCurrentCatalog())
-                        .listDatabases().stream()
-                        .sorted()
-                        .toArray(String[]::new);
-        return buildStringArrayResult("database name", databases);
+        String cName =
+                catalogName == null ? ctx.getCatalogManager().getCurrentCatalog() : catalogName;
+        List<String> databases =
+                ctx.getCatalogManager().getCatalogOrThrowException(cName).listDatabases();
+
+        if (likeType != null) {
+            databases =
+                    databases.stream()
+                            .filter(
+                                    row -> {
+                                        if (likeType == LikeType.ILIKE) {
+                                            return notLike
+                                                    != SqlLikeUtils.ilike(row, likePattern, "\\");
+                                        } else if (likeType == LikeType.LIKE) {
+                                            return notLike
+                                                    != SqlLikeUtils.like(row, likePattern, "\\");
+                                        }
+                                        return false;
+                                    })
+                            .collect(Collectors.toList());
+        }
+
+        return buildStringArrayResult(
+                "database name", databases.stream().sorted().toArray(String[]::new));
```

Comment:
Thanks for improving this!  I'll leave it up to @dawidwys if he has any other ideas here.  

I'm tempted to suggest something that doesn't collect the intermediate list.  (On one hand, this code isn't in an inner loop, so optimization may not be strictly necessary.  On another hand, it is useful to think through minimizing intermediate work!)

Reply:
If `databases` is a `Stream<String>` you wouldn't need to collect the filtering results.

```
Stream<String> databases =
                ctx.getCatalogManager().getCatalogOrThrowException(cName).listDatabases().stream();

        if (likeType != null) {
            databases =
                    databases
                            .filter(
                                    row -> {
                                        if (likeType == LikeType.ILIKE) {
                                            return notLike
                                                    != SqlLikeUtils.ilike(row, likePattern, "\\");
                                        } else if (likeType == LikeType.LIKE) {
                                            return notLike
                                                    != SqlLikeUtils.like(row, likePattern, "\\");
                                        }
                                        return false;
                                    });
        }
```